services/retention/service.go

```diff
@@ -80,83 +103,124 @@ func (s *Service) run() {
-						}
-					}
+
+			// Determine all shards that have expired and need to be deleted.
+			for _, g := range r.ExpiredShardGroups(time.Now().UTC()) {
+				if err := s.MetaClient.DeleteShardGroup(d.Name, r.Name, g.ID); err != nil {
+					log.Info("Failed to delete shard group",
+						logger.Database(d.Name),
+						logger.ShardGroup(g.ID),
+						logger.RetentionPolicy(r.Name),
+						zap.Error(err))
+					retryNeeded = true
+					continue
+				}
+
+				log.Info("Deleted shard group",
+					logger.Database(d.Name),
+					logger.ShardGroup(g.ID),
+					logger.RetentionPolicy(r.Name))
+
+				// Store all the shard IDs that may possibly need to be removed locally.
+				for _, sh := range g.Shards {
+					deletedShardIDs[sh.ID] = deletionInfo{db: d.Name, rp: r.Name}
 				}
 			}
+		}
+	}
 
-			// Remove shards if we store them locally
-			for _, id := range s.TSDBStore.ShardIDs() {
-				if info, ok := deletedShardIDs[id]; ok {
-					if err := s.TSDBStore.DeleteShard(id); err != nil {
-						log.Info("Failed to delete shard",
-							logger.Database(info.db),
-							logger.Shard(id),
-							logger.RetentionPolicy(info.rp),
-							zap.Error(err))
-						retryNeeded = true
-						continue
-					}
-					log.Info("Deleted shard",
+	// Remove shards if we store them locally
+	for _, id := range s.TSDBStore.ShardIDs() {
+		if info, ok := deletedShardIDs[id]; ok {
+			delete(deletedShardIDs, id)
+			if err := s.TSDBStore.DeleteShard(id); err != nil {
+				log.Error("Failed to delete shard",
+					logger.Database(info.db),
+					logger.Shard(id),
+					logger.RetentionPolicy(info.rp),
+					zap.Error(err))
+				if errors.Is(err, tsdb.ErrShardNotFound) {
+					log.Warn("Shard does not exist in store, continuing retention removal",
 						logger.Database(info.db),
 						logger.Shard(id),
 						logger.RetentionPolicy(info.rp))
+				} else {
+					retryNeeded = true
+					continue
 				}
 			}
-
-			if err := s.MetaClient.PruneShardGroups(); err != nil {
-				log.Info("Problem pruning shard groups", zap.Error(err))
+			log.Info("Deleted shard",
+				logger.Database(info.db),
+				logger.Shard(id),
+				logger.RetentionPolicy(info.rp))
+			if err := dropShardMetaRef(id, info); err != nil {
+				// removeShardMetaReference already logged the error.
 				retryNeeded = true
+				continue
 			}
+		}
+	}
 
-			if retryNeeded {
-				log.Info("One or more errors occurred during shard deletion and will be retried on the next check", logger.DurationLiteral("check_interval", time.Duration(s.config.CheckInterval)))
-			}
-
-			logEnd()
+	// Check for expired phantom shards that exist in the metadata but not in the store.
+	for id, info := range deletedShardIDs {
+		log.Error("Expired phantom shard detected during retention check, removing from metadata",
+			logger.Database(info.db),
+			logger.Shard(id),
+			logger.RetentionPolicy(info.rp))
+		if err := dropShardMetaRef(id, info); err != nil {
+			// removeShardMetaReference already logged the error.
```

Comment:
`dropShardMetaRef`

Reply:
`dropShardMetaRef` is passed in during service creation. It is parameterized because single-node OSS and clustered enterprise require different methods to drop the meta store reference.

The Enterprise PR uses `RemoveShardOwner` for `dropShardMetaRef`. This allows each node to drop it's reference to the shard when it deletes the TSM files from disk. Already existing code will implicitly drop the shard from the meta store once the shard owner count reaches 0. However, OSS does not make use of the shard owners field in the meta store. It's way of dropping the shard reference is to simply call `DropShard` after the TSM files have been removed from disk. Once all the shards in the shard group have been dropped the modified code in `PruneShardGroups` will then be able to remove the shard group.